File: app/src/main/java/com/dougkeen/bart/backend/HttpTransitFeedClient.java
package com.dougkeen.bart.backend;

import com.google.transit.realtime.GtfsRealtime;
import com.dougkeen.bart.networktasks.NetworkUtils;

import java.io.IOException;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

/** Downloads BART's complete trip-update and alert feeds. */
public final class HttpTransitFeedClient implements TransitFeedClient {
    public static final String TRIP_UPDATES_URL =
            "https://api.bart.gov/gtfsrt/tripupdate.aspx";
    public static final String ALERTS_URL =
            "https://api.bart.gov/gtfsrt/alerts.aspx";

    private final OkHttpClient client;

    public HttpTransitFeedClient() {
        this(NetworkUtils.makeHttpClient());
    }

    public HttpTransitFeedClient(OkHttpClient client) {
        this.client = client;
    }

    @Override
    public TransitFeedSnapshot fetch() throws IOException {
        GtfsRealtime.FeedMessage tripUpdates = fetchFeed(TRIP_UPDATES_URL);
        GtfsRealtime.FeedMessage alerts = fetchFeed(ALERTS_URL);
        return new TransitFeedSnapshot(tripUpdates, alerts,
                System.currentTimeMillis());
    }

    private GtfsRealtime.FeedMessage fetchFeed(String url) throws IOException {
        Request request = new Request.Builder()
                .url(url)
                .header("Accept", "application/x-google-protobuf")
                .build();
        try (Response response = client.newCall(request).execute()) {
            if (!response.isSuccessful()) {
                throw new IOException("Realtime feed returned "
                        + response.code() + " for " + url);
            }
            if (response.body() == null) {
                throw new IOException("Realtime feed returned an empty body for "
                        + url);
            }
            return GtfsRealtime.FeedMessage.parseFrom(
                    response.body().byteStream());
        }
    }
}

File: app/src/main/kotlin/com/dougkeen/bart/backend/HttpTransitFeedClient.kt
package com.dougkeen.bart.backend

import com.dougkeen.bart.networktasks.NetworkUtils
import com.google.transit.realtime.GtfsRealtime
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.IOException

/** Downloads BART's complete trip-update and alert feeds. */
class HttpTransitFeedClient : TransitFeedClient {
    @JvmOverloads
    constructor(client: OkHttpClient = NetworkUtils.makeHttpClient()) {
        this.client = client
    }

    private val client: OkHttpClient

    override fun fetch(): TransitFeedSnapshot {
        val tripUpdates = fetchFeed(TRIP_UPDATES_URL)
        val alerts = fetchFeed(ALERTS_URL)
        return TransitFeedSnapshot(tripUpdates, alerts, System.currentTimeMillis())
    }

    override fun fetchFeeds(): TransitFeedFetchResult {
        var tripUpdates: GtfsRealtime.FeedMessage? = null
        var alerts: GtfsRealtime.FeedMessage? = null
        var tripUpdatesError: Exception? = null
        var alertsError: Exception? = null
        try {
            tripUpdates = fetchFeed(TRIP_UPDATES_URL)
        } catch (exception: Exception) {
            tripUpdatesError = exception
        }
        try {
            alerts = fetchFeed(ALERTS_URL)
        } catch (exception: Exception) {
            alertsError = exception
        }
        return TransitFeedFetchResult(
            tripUpdates,
            tripUpdatesError,
            alerts,
            alertsError
        )
    }

    private fun fetchFeed(url: String): GtfsRealtime.FeedMessage {
        val request = Request.Builder()
            .url(url)
            .header("Accept", "application/x-google-protobuf")
            .build()
        return client.newCall(request).execute().use { response ->
            if (!response.isSuccessful) {
                throw IOException("Realtime feed returned ${response.code} for $url")
            }
            val body = response.body
                ?: throw IOException("Realtime feed returned an empty body for $url")
            GtfsRealtime.FeedMessage.parseFrom(body.byteStream())
        }
    }

    companion object {
        @JvmField
        val TRIP_UPDATES_URL = "https://api.bart.gov/gtfsrt/tripupdate.aspx"

        @JvmField
        val ALERTS_URL = "https://api.bart.gov/gtfsrt/alerts.aspx"
    }
}
